cpp/src/arrow/compute/kernels/scalar_cast_extension.cc

```diff
@@ -0,0 +1,74 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Implementation of casting to extension types
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/kernels/scalar_cast_internal.h"
+#include "arrow/scalar.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+namespace {
+Status CastToExtension(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+  const CastOptions& options = checked_cast<const CastState*>(ctx->state())->options;
+  auto out_ty = static_cast<const ExtensionType&>(*options.to_type.type).storage_type();
+
+  DCHECK(batch[0].is_array());
+  std::shared_ptr<Array> array = batch[0].array.ToArray();
+
+  // Try to prevent user errors by preventing casting between extensions w/
+  // different storage types. Provide a tip on how to accomplish same outcome.
+  std::shared_ptr<Array> result;
+  if (array->type()->id() == Type::EXTENSION) {
+    if (!array->type()->Equals(out_ty)) {
+      return Status::Invalid("Casting from '" + array->type()->ToString() +
+                             "' to extension with different storage type '" +
```

Comment:
```suggestion
                             "' to extension with different type '" +
```

It's just a different type that is not allowed, right? And not only different storage type?

Reply:
Done and done. :+1: